libutils/passopenfile.c

```diff
@@ -0,0 +1,509 @@
+                        {
+                            Log(LOG_LEVEL_ERR,
+                                "Failed to receive whole text accompanying descriptor");
+                        }
+                    }
+                    else
+                    {
+                        got = 0; /* != size */
+                        Log(LOG_LEVEL_ERR,
+                            "Failed to allocate buffer for text accompanying descriptor");
+                    }
+
+                    if (text && got == size)
+                    {
+                        *text = buffer;
+                    }
+                    else
+                    {
+                        free(buffer);
+                    }
+                }
+            }
+            else
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Failed to read size of text accompanying descriptor");
+                size = 0;
+            }
+
+            if (text && size && !*text)
+            {
+                closesocket(descriptor);
+                return -1;
+            }
+        }
+        else
+        {
+            Log(LOG_LEVEL_ERR,
+                "Failed to receive whole descriptor blob");
+            return -1;
+        }
+    }
+
+    /* Send ACK now that we've opened our end of the socket. */
+    if (!wait_for(uds, true, &ready))
+    {
+        Log(LOG_LEVEL_ERR,
+            "Can't ACK received descriptor (select: %s)",
+            GetErrorStr());
+    }
+    else if (!ready)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "No descriptor supplier to ACK to, aborting receipt.");
+    }
+    else
+    {
+        send(uds, ACK_MSG, sizeof(ACK_MSG), 0);
+
+        return descriptor;
+    }
+
+    closesocket(descriptor);
+    return -1;
+}
+
+#else /* Unix: */
+
+#ifdef HAVE_MSGHDR_MSG_CONTROL
+/* This is the modern interface, should be present in most modern
+ * Linux systems.  Transferred file descriptors "behave as though they
+ * have been created with dup(2)" according to unix(7); which means
+ * the sender can close() its end without prejudice to the recipient's
+ * use of its copy. */
+#define INTERFACE_STYLE "Linux"
+#elif HAVE_MSGHDR_ACCRIGHTS
+/* This is the old (BSD) interface.  No Linux should use this, it's
+ * only known to be present in BSD-based Unixes. */
+#define INTERFACE_STYLE "BSD"
+#else
+#error "No support for connection sharing, this code should not be compiled"
+#endif
+
+#ifndef MSG_WAITALL /* Linux >= 2.2 */
+#define MSG_WAITALL 0
+#endif
+static const char NULL_MSG[] = "\0NULL";
+
+bool PassOpenFile_Put(int uds, int descriptor, const char *text)
+{
+    struct msghdr message;
+    struct iovec vector;
+    size_t msglen = text ? strlen(text) + 1 : sizeof(NULL_MSG);
+    assert(MAX_MESSAGE_SIZE >= msglen);
+    Log(LOG_LEVEL_VERBOSE,
+        "Connected to peer, passing descriptor %d with %s %s",
+        descriptor,
+        text ? "text:" : "no",
+        text ? text : "text");
```

Comment:
I think this should be debug log level, it will be quite verbose since this happens on every connection, right? It's also too technical to be useful for the user.